src/wovnio/wovnphp/custom_domain/CustomDomainLangs.php

```diff
@@ -43,6 +65,44 @@ public function getCustomDomainLangByUrl($url)
         return count($results) <= 0 ? null : array_shift($results);
     }
 
+    public function toHtmlSwapperHash()
+    {
+        $result = array();
+        foreach ($this->customDomainLangs as $lang) {
+            $result[$lang->getHostAndPathWithoutTrailingSlash()] = $lang->getLang();
+        }
+        return $result;
+    }
+
+    /**
+     * Returns the computed (virtual) uri representation for a given physical location uri.
+     * Used when communicating with html-swapper.
+     *
+     * @param $uri string the current uri - pointing to physical location of current lang
+     * @param $lang string lang code of the current uri
+     * @param $defaultLang string lang code of the default (source) language
+     * @return string|string[]|null
+     */
+    public function computeSourceVirtualUrl($uri, $lang, $defaultLang)
+    {
+        $currentLangDomainLang = $this->getSourceCustomDomainByLang($lang);
+        if ($currentLangDomainLang->getSource()) {
+            $defaultCustomDomainLang = $currentLangDomainLang->getSource();
+        } else {
+            $defaultCustomDomainLang = $this->getCustomDomainLangByLang($defaultLang);
+        }
+        return CustomDomainLangUrlHandler::changeToNewCustomDomainLang($uri, $currentLangDomainLang, $defaultCustomDomainLang);
+    }
+
+    public function hasSource($langCode)
```

Comment:
Ideally I would like to remove this function.

Reply:
We must test for if a custom domain have a source set, otherwise will just convert "no lang" url back to ones that has lang.